src/io/SeltzerBergerReader.cc

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file SeltzerBergerReader.cc
+//---------------------------------------------------------------------------//
+#include "SeltzerBergerReader.hh"
+
+#include <fstream>
+#include <sstream>
+#include "base/Assert.hh"
+#include "base/Range.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct using environmental variable $G4LEDATA.
+ */
+SeltzerBergerReader::SeltzerBergerReader()
+{
+    const char* env_var = std::getenv("G4LEDATA");
+    CELER_VALIDATE(env_var, "Environment variable G4LEDATA is not defined.");
+    std::ostringstream os;
+    os << env_var << "/brem_SB";
+    path_ = os.str();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct using a user defined path to the folder containing the data.
+ * The path should point to the files that are usually stored in
+ * [Geant4-install]/share/Geant4-10.7.0/data/G4EMLOW7.12/brem_SB/.
+ */
+SeltzerBergerReader::SeltzerBergerReader(const char* path) : path_(path)
+{
+    CELER_EXPECT(path_.size());
+    if (path_.back() == '/')
+    {
+        path_.pop_back();
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Fetch data for a given atomic number.
+ *
+ * Standard data files encompass Z = [1, 100].
+ */
+SeltzerBergerReader::result_type
+SeltzerBergerReader::operator()(AtomicNumber atomic_number) const
+{
+    CELER_EXPECT(atomic_number > 0);
+
+    // For Z = 93-99, the incident log energy grid and reduced photon energy
+    // grid in the bremsstrahlung data files are incorrect. These are the grids
+    // that should be used for those elements (and for all Z < 100).
+    constexpr double log_energy[]
+        = {-6.9078,  -6.5023,  -6.2146,  -5.8091, -5.5215, -5.2983, -5.116,
+           -4.8283,  -4.6052,  -4.1997,  -3.912,  -3.5066, -3.2189, -2.9957,
+           -2.8134,  -2.5257,  -2.3026,  -1.8971, -1.6094, -1.204,  -0.91629,
+           -0.69315, -0.51083, -0.22314, 0,       0.40547, 0.69315, 1.0986,
+           1.3863,   1.6094,   1.7918,   2.0794,  2.3026,  2.7081,  2.9957,
+           3.4012,   3.6889,   3.912,    4.0943,  4.382,   4.6052,  5.0106,
+           5.2983,   5.7038,   5.9915,   6.2146,  6.3969,  6.6846,  6.9078,
+           7.3132,   7.6009,   8.0064,   8.294,   8.5172,  8.6995,  8.9872,
+           9.2103};
+    constexpr double kappa[]
+        = {1e-12, 0.025, 0.05,  0.075,  0.1,    0.15,    0.2,     0.25,
+           0.3,   0.35,  0.4,   0.45,   0.5,    0.55,    0.6,     0.65,
+           0.7,   0.75,  0.8,   0.85,   0.9,    0.925,   0.95,    0.97,
+           0.99,  0.995, 0.999, 0.9995, 0.9999, 0.99995, 0.99999, 1.};
```

Comment:
@sethrj @amandalund I am not sure whether I understand this issue as I do not see anything wrong to have different energy/kappa points (i.e., two different sets, one up to Z=92 and the other in Z=[93-100]) of the brem_sb table - these points are parameterized values by Seltzer and Berger - I guess that original data size are statistically small at the high Z points and they parameterized data and prepared tables with relatively smaller number of fit points than low Z data .  As the data table are retrieved based on the energy/kappa bin numbers specified in the first line, there may be no issue as long as they are used consistently.  The SB sampling (alias) tables has a different story as they were re-parameterized based on the original brem_sb tables and use the same grid convension.

Reply:
@whokion there's no issue with having different energy grids for different Z; the problem is that the energy grids for Z=93-100 don't match the DCS values for those elements. The dimension of the grids is 14x31, but the dimension of the data is 32x57 -- so if you retrieve the data based on the bin numbers provided, you'll only read in the first 434 values, and they won't correspond to the correct energy grid values.

Reply:
@amandalund I thought that we agreed to use the brem_SB tables instead of DCS tables (i.e., 10.4 sampling procedure for now, using the composition and rejection method instead of the alias sampling).  Again DCS tables are derived from bream_SB (i.e, equivalently differential cross sections) and can have a different grid convention (i.e., using the same 32x57 points for every Z for DCS since 10.5).

Reply:
We are using the 10.4 sampling procedure  -- the data tables there *are* labeled as differential cross sections, although I think they're scaled by 1/x .

Reply:
Go to `$G4LEDATA/brem_SB` and look at the difference between br92 and br93. These are what we're loading. The tabulated data is essentially the same (within a  few percent) but the grids sizes and values on lines 1–3 are completely wrong, inconsistent with the tables below.

Reply:
@sethrj @amandalund  Okay, I see the problem.  It seems that this is a real bug (if a Geant4 application uses any element of which Z > 92 for electron Bremsstrahlung.  Probably those elements have never been simulated at least for any HEP detector, at least in US).  The right set of bins should be 32x57 too for those tables of which Z > 92 based on cross section values of Z=92 - those tables for Z > 92 (and Z=1) are extrapolated parameterization based on data of Z = 2, 8, 13, 47, 79, and 92 (see original data table of which data points are the 14x31 grid shown in "Nuclear Instruments and Methods in Physics Research B12 (1985) 95-134". I will double check the issue with Vladimir Ivantchenko (Geant4 EM group leader).

Reply:
Thanks @whokion ! @amandalund i hate t do this to you, but maybe since we don’t need these data, we should skip the hard coded workaround and instead add a validate at the end of the element import that we’ve actually reached the end of file? That should catch this or similar errors.

Reply:
Thanks @whokion! @sethrj sure thing 🙂